### PR TITLE
Frogger initial commit

### PR DESCRIPTION
**Pull request by yaripey**
### Frogger

[Demo](https://yaripey.github.io/frontend-nanodegree-arcade-game/)
[Code base](https://github.com/yaripey/frontend-nanodegree-arcade-game)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

**Comment by andrewklmn** [comment]
@yaripey 
Good job!
All bugs are running and frogger is moving well!

no_difference
But there is no difference if I lose or win.
In all these cases I just return to the start position.
Maybe it's better to add an alert() with congratulations when I win?

It's not a code review. Just wondering.

**Comment by yaripey** [comment]
Yeah, I like the idea. Will implement.

File: submissions/yaripey/frogger/app.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    this.x = x;
+    this.y = y;
+    this.changeSpeed();
+    // The image/sprite for our enemies, this uses
+    // a helper we've provided to easily load images
+    this.sprite = 'images/enemy-bug.png';
+};
+
+Enemy.prototype.changeSpeed = function () {
+    this.speed = Math.random() * (400 - 100) + 100;
```
Comment (andrewklmn) [inline submissions/yaripey/frogger/app.js:20]: magic numbers...
All numbers like block dimensions, initial locations should be defined as constants.

File: submissions/yaripey/frogger/app.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+// Update the enemy's position, required method for game
+// Parameter: dt, a time delta between ticks
+Enemy.prototype.update = function (dt) {
+    if (this.x > tile.width * 5) {
```
Comment (andrewklmn) [inline submissions/yaripey/frogger/app.js:26]: magic number '5'....

File: submissions/yaripey/frogger/app.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        this.changeSpeed()
+        this.x = -tile.width * 1
```
Comment (andrewklmn) [inline submissions/yaripey/frogger/app.js:28]: do we really need this multiplication by 1 here?

File: submissions/yaripey/frogger/app.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+// Now instantiate your objects.
+// Place all enemy objects in an array called allEnemies
+// Place the player object in a variable called player
+const enemy1 = new Enemy(-tile.width * 2, tile.height * 0.5)
+const enemy2 = new Enemy(-tile.width * 2, tile.height * 1.5)
+const enemy3 = new Enemy(-tile.width * 2, tile.height * 2.5)
+const allEnemies = [enemy1, enemy2, enemy3]
+const player = new Player(tile.width * 2, tile.height * 4.5)
```
Comment (andrewklmn) [inline submissions/yaripey/frogger/app.js:92]: initial locations should be defined as constants

File: submissions/yaripey/frogger/app.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+const checkCollisions = () => {
+    if (player.y < 0) {
+        alert('You\'ve made it to the water! Congratulations!')
+        player.resetPosition()
+    }
+    if (allEnemies.some(enemy => {
+        if (
+            Math.abs(Math.abs(enemy.x) - Math.abs(player.x)) < 60
+            && Math.abs(Math.abs(enemy.y) - Math.abs(player.y)) < 60
```
Comment (andrewklmn) [inline submissions/yaripey/frogger/app.js:116]: magic number '60'...
Is it COLLISION_DISTANCE? maybe it's better to use a defined constant instead? e.g.:
const COLLISION_DISTANCE = 60;

File: submissions/yaripey/frogger/app.js [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+// Enemies our player must avoid
+const Enemy = function (x, y) {
+    // Variables applied to each of our instances go here,
+    // we've provided one for you to get started
+    this.x = x;
+    this.y = y;
+    this.changeSpeed();
+    // The image/sprite for our enemies, this uses
+    // a helper we've provided to easily load images
+    this.sprite = 'images/enemy-bug.png';
```
Comment (andrewklmn) [inline submissions/yaripey/frogger/app.js:16]: this.x, this.y, this.sprite - properties that can be generalized into a base class. For example, the class 'Character' can be a base class, which is extended by Enemy and Player classes. And you can use inheritance for them both.

Class extension can be implemented using Subclass.prototype = Object.create(Superclass.prototype);
[Here is useful article](https://ui.dev/javascript-inheritance-and-the-prototype-chain/) about functional inheritance.

File: submissions/yaripey/frogger/app.js [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+// a handleInput() method.
+const Player = function (x, y) {
+    this.initialX = x;
+    this.initialY = y;
+    this.x = x;
+    this.y = y;
+    this.stepX = tile.width;
+    this.stepY = tile.height;
+
+    this.sprite = "images/char-boy.png";
```
Comment (andrewklmn) [inline submissions/yaripey/frogger/app.js:52]: Definition of this.x, this.y, this.sprite - in 'Player' also can be moved to the base 'Character' class...

File: submissions/yaripey/frogger/app.js [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+Player.prototype.update = function (dt) {
+
+}
+
+Player.prototype.render = function () {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+}
```
Comment (andrewklmn) [inline submissions/yaripey/frogger/app.js:61]: Method 'render()' of 'Player' class is the same as method 'render()' of Enemy... 
So we can avoid this repetition of code just by moving this method to the prototype of the base 'Character' class.

**Review by andrewklmn — CHANGES_REQUESTED** [review]
Good job!
Your code is well-formatted and easy to read. Thank you!

Let's make this good code more object-oriented!
Check my comments below and feel free to ask any questions.

**Comment by andrewklmn** [comment]
> Yeah, I like the idea. Will implement.

@yaripey 
Well done!
It works now! )))

well_done

**Review by andrewklmn — APPROVED** [review]
@yaripey 
LGTM!
Good job!
Now your code looks more DRYer!...
I approve.

**Comment by andrewklmn** [comment]
@A-Ostrovnyy 
@OleksiyRudenko 
@zonzujiro 

Could you check approve and merge this PR, please?

**Pull request merged**